{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dedecad5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import dill\n",
    "import scanpy as sc\n",
    "import os\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# import pyranges\n",
    "# Set stderr to null to avoid strange messages from ray\n",
    "import sys\n",
    "_stderr = sys.stderr\n",
    "null = open(os.devnull,'wb')\n",
    "from scipy.sparse import issparse\n",
    "\n",
    "\n",
    "import itertools\n",
    "import anndata\n",
    "import itertools\n",
    "import anndata\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt \n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore')\n",
    "import numpy as np\n",
    "from sklearn.preprocessing import normalize\n",
    "\n",
    "from itertools import product\n",
    "import umap\n",
    "from collections import Counter\n",
    "from sklearn.decomposition import PCA\n",
    "from gridot.utils import *\n",
    "from gridot.train import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c06b6b0d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 2664 × 21255\n",
      "    obs: 'n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'ingest_celltype_label', 'leiden_res_0.8', 'celltype'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells', 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts'\n"
     ]
    }
   ],
   "source": [
    "rna_anndata = sc.read_h5ad('./data/adata.h5ad')\n",
    "\n",
    "# Check if raw data exists\n",
    "if rna_anndata.raw is not None:\n",
    "    # Access the raw data\n",
    "    raw_data = rna_anndata.raw.X\n",
    "    raw_obs = rna_anndata.obs.copy()  # Cell metadata (unchanged)\n",
    "    raw_var = rna_anndata.raw.var.copy()  # Gene metadata (original raw genes)\n",
    "\n",
    "    # Create a new AnnData object using the raw data\n",
    "    rna_anndata = sc.AnnData(X=raw_data, obs=raw_obs, var=raw_var)\n",
    "\n",
    "    # Print the new raw AnnData object to verify\n",
    "    print(rna_anndata)\n",
    "else:\n",
    "    print(\"The AnnData object does not contain raw data.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "105a23a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "         B_cells CD4_T_cells CD8_T_cells CD14+_Monocytes Dendritic_cells  \\\n",
      "0           AFF3      BCL11B        CCL5          PLXDC2            CD74   \n",
      "1           CD74       CAMK4        NKG7            FCN1         HLA-DRA   \n",
      "2        HLA-DRA        LEF1        GZMA            DPYD            CST3   \n",
      "3          BANK1        IL7R        IL32          SLC8A1        HLA-DRB1   \n",
      "4        RALGPS2        TCF7       PRKCH           LRMDA        HLA-DPA1   \n",
      "...          ...         ...         ...             ...             ...   \n",
      "5308  AC124312.5       SOCS5     SNRNP27        WWC2-AS2      AL357079.1   \n",
      "5309  AL121890.4      SEC22A      FER1L6      AL121987.2         CCDC146   \n",
      "5310       WDR64     CARHSP1     SPINDOC            ELDR          LRRC72   \n",
      "5311      UBE2J2       ACKR3       ACOT2         SULT1A4       LINC02021   \n",
      "5312       BAMBI       NDST4  AC013437.1           CNTN4          ALKBH4   \n",
      "\n",
      "     FCGR3A+_Monocytes    NK_cells  \n",
      "0                 PSAP        GNLY  \n",
      "1                 CTSS        NKG7  \n",
      "2                 WARS        PRF1  \n",
      "3                 CST3       CD247  \n",
      "4               FCGR3A        GZMA  \n",
      "...                ...         ...  \n",
      "5308            GTPBP6  AP000919.3  \n",
      "5309             LYRM9       TTC26  \n",
      "5310            NDUFA2  AC005014.4  \n",
      "5311           SLC30A1        TTC1  \n",
      "5312        AC097534.2        DRC3  \n",
      "\n",
      "[5313 rows x 7 columns]\n"
     ]
    }
   ],
   "source": [
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Assuming your AnnData object is named `rna_anndata` and has a 'cell_type' column in .obs\n",
    "groupby_key = 'celltype'\n",
    "\n",
    "# Step 1: Perform differential expression\n",
    "sc.tl.rank_genes_groups(rna_anndata, groupby=groupby_key, method='t-test')  # You can use 'wilcoxon' too\n",
    "\n",
    "# Step 2: Extract DE results\n",
    "de_results = rna_anndata.uns['rank_genes_groups']\n",
    "groups = de_results['names'].dtype.names  # list of cell type names\n",
    "\n",
    "top_genes_dict = {}\n",
    "\n",
    "# Step 3: For each group (cell type), get top 25% DE genes\n",
    "for group in groups:\n",
    "    gene_names = np.array(de_results['names'][group])  # ranked gene names\n",
    "    n_top = int(len(gene_names) * 0.25)  # top 25%\n",
    "    top_genes = gene_names[:n_top]\n",
    "    top_genes_dict[group] = top_genes\n",
    "\n",
    "# Optional: Convert to DataFrame for easy inspection\n",
    "top_genes_df = pd.DataFrame(dict([(k, pd.Series(v)) for k, v in top_genes_dict.items()]))\n",
    "\n",
    "# Display or save\n",
    "print(top_genes_df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "38b12703",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "putative_pairs = pd.read_csv(f\"./testing_results_pseudo/TF_motif/putative_TF_gene.tsv\", sep='\\t', index_col=0)\n",
    "\n",
    "# Dictionary to store filtered pairs per cell type\n",
    "filtered_pairs_by_celltype = {}\n",
    "\n",
    "for cell_type, de_genes in top_genes_dict.items():\n",
    "    # Convert list to set for faster lookup\n",
    "    de_genes_set = set(de_genes)\n",
    "\n",
    "    # Filter putative_pairs where Target is in DE gene list for that cell type\n",
    "    filtered_pairs = putative_pairs[putative_pairs['TF'].isin(de_genes_set)]\n",
    "\n",
    "    # Add extra columns\n",
    "    filtered_pairs['rss_ratio'] = 1\n",
    "    filtered_pairs['ftest'] = 1\n",
    "\n",
    "    # Save to dictionary AFTER adding new columns\n",
    "    filtered_pairs_by_celltype[cell_type] = filtered_pairs\n",
    "\n",
    "    # Save each to file\n",
    "    filtered_pairs.to_csv(f\"./testing_results_pseudo/filtered_pairs_{cell_type}.tsv\", sep='\\t', index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0aecf043",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TF</th>\n",
       "      <th>Gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZNF251</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FOXI1</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATF1</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CREB3L4</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>PAX3</td>\n",
       "      <td>A1BG-AS1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1880495</th>\n",
       "      <td>SALL2</td>\n",
       "      <td>CWC25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1880503</th>\n",
       "      <td>XBP1</td>\n",
       "      <td>CWC25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1880506</th>\n",
       "      <td>TFE3</td>\n",
       "      <td>CWC25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1880524</th>\n",
       "      <td>NR2E1</td>\n",
       "      <td>CWC25</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1880531</th>\n",
       "      <td>LHX5</td>\n",
       "      <td>CWC25</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>996749 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TF      Gene\n",
       "0         ZNF251  A1BG-AS1\n",
       "1          FOXI1  A1BG-AS1\n",
       "2           ATF1  A1BG-AS1\n",
       "3        CREB3L4  A1BG-AS1\n",
       "4           PAX3  A1BG-AS1\n",
       "...          ...       ...\n",
       "1880495    SALL2     CWC25\n",
       "1880503     XBP1     CWC25\n",
       "1880506     TFE3     CWC25\n",
       "1880524    NR2E1     CWC25\n",
       "1880531     LHX5     CWC25\n",
       "\n",
       "[996749 rows x 2 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "putative_pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "879e44fb",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gridot",
   "language": "python",
   "name": "gridot"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
