{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "20e0026f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import dill\n",
    "import scanpy as sc\n",
    "import os\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# import pyranges\n",
    "# Set stderr to null to avoid strange messages from ray\n",
    "import sys\n",
    "_stderr = sys.stderr\n",
    "null = open(os.devnull,'wb')\n",
    "from scipy.sparse import issparse\n",
    "\n",
    "\n",
    "import itertools\n",
    "import anndata\n",
    "import itertools\n",
    "import anndata\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt \n",
    "import warnings\n",
    "warnings.simplefilter(action='ignore')\n",
    "import numpy as np\n",
    "from sklearn.preprocessing import normalize\n",
    "\n",
    "from itertools import product\n",
    "import umap\n",
    "from collections import Counter\n",
    "from sklearn.decomposition import PCA\n",
    "from gridot.utils import *\n",
    "from gridot.train import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4537241b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2664 × 104227\n",
       "    obs: 'celltype', 'n_genes', 'orig.ident'\n",
       "    var: 'Chromosome', 'Start', 'End', 'n_cells', 'features', 'chr_no', 'start', 'end'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac_anndata = sc.read_h5ad('./data/pseudo_atac_pbmc_binarized_20_20.h5ad')\n",
    "atac_anndata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "90587cc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "atac_anndata.var['chr_no'] = atac_anndata.var['Chromosome']\n",
    "atac_anndata.var['start'] = atac_anndata.var['Start']\n",
    "atac_anndata.var['end'] = atac_anndata.var['End']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bb0b619",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "4de0a1bf",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 2664 × 21255\n",
      "    obs: 'n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'ingest_celltype_label', 'leiden_res_0.8', 'celltype'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells', 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts'\n"
     ]
    }
   ],
   "source": [
    "rna_anndata = sc.read_h5ad('./data/adata.h5ad')\n",
    "\n",
    "# Check if raw data exists\n",
    "if rna_anndata.raw is not None:\n",
    "    # Access the raw data\n",
    "    raw_data = rna_anndata.raw.X\n",
    "    raw_obs = rna_anndata.obs.copy()  # Cell metadata (unchanged)\n",
    "    raw_var = rna_anndata.raw.var.copy()  # Gene metadata (original raw genes)\n",
    "\n",
    "    # Create a new AnnData object using the raw data\n",
    "    rna_anndata = sc.AnnData(X=raw_data, obs=raw_obs, var=raw_var)\n",
    "\n",
    "    # Print the new raw AnnData object to verify\n",
    "    print(rna_anndata)\n",
    "else:\n",
    "    print(\"The AnnData object does not contain raw data.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a3968e8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltypes = rna_anndata.obs['celltype'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "57ffd90c",
   "metadata": {},
   "outputs": [],
   "source": [
    "schema_reference = 'rna'\n",
    "marker_gene = 'CD38'\n",
    "distance_threshold = 1e5\n",
    "device = 'cuda:0'\n",
    "num_components = 50\n",
    "gene_file = \"./data/gencode.v38.annotation.gff3\"\n",
    "folder_path = './testing_results'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "02c21322",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pre-processing scATAC-seq...\n",
      "AnnData object with n_obs × n_vars = 2664 × 104227\n",
      "    obs: 'celltype', 'n_genes', 'orig.ident'\n",
      "    var: 'Chromosome', 'Start', 'End', 'n_cells', 'features', 'chr_no', 'start', 'end'\n",
      "Pre-processing scRNA-seq...\n",
      "AnnData object with n_obs × n_vars = 2664 × 21167\n",
      "    obs: 'n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'ingest_celltype_label', 'leiden_res_0.8', 'celltype'\n",
      "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells', 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'features'\n",
      "Adding start site to meta file...\n",
      "Processing multimodal\n",
      "CRE-gene relationship\n",
      "Identifying candidate pairs within 100000.0 bp\n",
      "244140\n",
      "Running quadratic program... done.\n",
      "Cell with highest CD38 expression: ('CTAGCTTGTTGAGGTC-1', 1124)\n",
      "Expression level:   (0, 0)\t5.438936233520508\n",
      "WARNING: Trying to run `tl.dpt` without prior call of `tl.diffmap`. Falling back to `tl.diffmap` with default parameters.\n",
      "Epoch 0 (18.48 seconds): train loss 1016.12\n",
      "Epoch Inference (30.78 seconds): test loss 988.70\n",
      "Total Time: 49.721131801605225 seconds\n",
      "Folder created at: ./testing_results\n"
     ]
    }
   ],
   "source": [
    "atac_adata, rna_adata = preprocessing_adata(atac_anndata,rna_anndata,gene_file=gene_file)\n",
    "\n",
    "print(\"Processing multimodal\")\n",
    "preprocess_multimodal(rna_adata,atac_adata)\n",
    "print(\"CRE-gene relationship\")\n",
    "run_gridot_cell_population(rna_adata,atac_adata,schema_reference='rna',marker_gene = marker_gene,distance_threshold=distance_threshold,device = device,folder_path = folder_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "430df1c4",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'chr_no'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m/oscar/data/rsingh47/tpham43/gridot_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:3800\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3799\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3800\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3801\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32m/oscar/data/rsingh47/tpham43/gridot_env/lib/python3.11/site-packages/pandas/_libs/index.pyx:138\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32m/oscar/data/rsingh47/tpham43/gridot_env/lib/python3.11/site-packages/pandas/_libs/index.pyx:165\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5745\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:5753\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'chr_no'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[11], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m candidates_df \u001b[38;5;241m=\u001b[39m \u001b[43midentify_all_peak_gene_link_candidates\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrna_anndata\u001b[49m\u001b[43m,\u001b[49m\u001b[43matac_anndata\u001b[49m\u001b[43m,\u001b[49m\u001b[43mdistance_thresh\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1e6\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m candidates_df\n",
      "File \u001b[0;32m/oscar/data/rsingh47/tpham43/GRIDOT/src/gridot/utils.py:296\u001b[0m, in \u001b[0;36midentify_all_peak_gene_link_candidates\u001b[0;34m(rna_adata, atac_adata, distance_thresh, rna_chr_key, rna_txstart_key, atac_chr_key, atac_start_key, atac_end_key)\u001b[0m\n\u001b[1;32m    293\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m chr_no \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28msorted\u001b[39m(\u001b[38;5;28mlist\u001b[39m(\u001b[38;5;28mset\u001b[39m(atac_adata\u001b[38;5;241m.\u001b[39mvar[atac_chr_key]\u001b[38;5;241m.\u001b[39mvalues))):\n\u001b[1;32m    295\u001b[0m \tfiltered_atac_df \u001b[38;5;241m=\u001b[39m atac_adata\u001b[38;5;241m.\u001b[39mvar[atac_adata\u001b[38;5;241m.\u001b[39mvar[atac_chr_key] \u001b[38;5;241m==\u001b[39m chr_no]\n\u001b[0;32m--> 296\u001b[0m \tfiltered_gene_df \u001b[38;5;241m=\u001b[39m rna_adata\u001b[38;5;241m.\u001b[39mvar[\u001b[43mrna_adata\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mvar\u001b[49m\u001b[43m[\u001b[49m\u001b[43mrna_chr_key\u001b[49m\u001b[43m]\u001b[49m \u001b[38;5;241m==\u001b[39m chr_no]\n\u001b[1;32m    297\u001b[0m \tfiltered_gene_symbols \u001b[38;5;241m=\u001b[39m filtered_gene_df\u001b[38;5;241m.\u001b[39mindex\u001b[38;5;241m.\u001b[39mvalues\n\u001b[1;32m    299\u001b[0m \tpeak_data \u001b[38;5;241m=\u001b[39m filtered_atac_df[[atac_start_key,atac_end_key]]\u001b[38;5;241m.\u001b[39mvalues\n",
      "File \u001b[0;32m/oscar/data/rsingh47/tpham43/gridot_env/lib/python3.11/site-packages/pandas/core/frame.py:3805\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   3804\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 3805\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   3807\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m/oscar/data/rsingh47/tpham43/gridot_env/lib/python3.11/site-packages/pandas/core/indexes/base.py:3802\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3800\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[1;32m   3801\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n\u001b[0;32m-> 3802\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3803\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3804\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3805\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3806\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'chr_no'"
     ]
    }
   ],
   "source": [
    "candidates_df = identify_all_peak_gene_link_candidates(rna_anndata,atac_anndata,distance_thresh=1e6)\n",
    "candidates_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "87a85643",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>features</th>\n",
       "      <th>chr_no</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>n_cells</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>chr2:148881654-148881927</th>\n",
       "      <td>chr2:148881654-148881927</td>\n",
       "      <td>chr2:148881654-148881927</td>\n",
       "      <td>2</td>\n",
       "      <td>148881654</td>\n",
       "      <td>148881927</td>\n",
       "      <td>chr2</td>\n",
       "      <td>148881654</td>\n",
       "      <td>148881927</td>\n",
       "      <td>93</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr8:99495798-99496023</th>\n",
       "      <td>chr8:99495798-99496023</td>\n",
       "      <td>chr8:99495798-99496023</td>\n",
       "      <td>8</td>\n",
       "      <td>99495798</td>\n",
       "      <td>99496023</td>\n",
       "      <td>chr8</td>\n",
       "      <td>99495798</td>\n",
       "      <td>99496023</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr22:21735444-21736416</th>\n",
       "      <td>chr22:21735444-21736416</td>\n",
       "      <td>chr22:21735444-21736416</td>\n",
       "      <td>22</td>\n",
       "      <td>21735444</td>\n",
       "      <td>21736416</td>\n",
       "      <td>chr22</td>\n",
       "      <td>21735444</td>\n",
       "      <td>21736416</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr3:28241304-28242015</th>\n",
       "      <td>chr3:28241304-28242015</td>\n",
       "      <td>chr3:28241304-28242015</td>\n",
       "      <td>3</td>\n",
       "      <td>28241304</td>\n",
       "      <td>28242015</td>\n",
       "      <td>chr3</td>\n",
       "      <td>28241304</td>\n",
       "      <td>28242015</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr5:154627638-154628386</th>\n",
       "      <td>chr5:154627638-154628386</td>\n",
       "      <td>chr5:154627638-154628386</td>\n",
       "      <td>5</td>\n",
       "      <td>154627638</td>\n",
       "      <td>154628386</td>\n",
       "      <td>chr5</td>\n",
       "      <td>154627638</td>\n",
       "      <td>154628386</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr15:71448618-71449096</th>\n",
       "      <td>chr15:71448618-71449096</td>\n",
       "      <td>chr15:71448618-71449096</td>\n",
       "      <td>15</td>\n",
       "      <td>71448618</td>\n",
       "      <td>71449096</td>\n",
       "      <td>chr15</td>\n",
       "      <td>71448618</td>\n",
       "      <td>71449096</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr15:71792584-71792805</th>\n",
       "      <td>chr15:71792584-71792805</td>\n",
       "      <td>chr15:71792584-71792805</td>\n",
       "      <td>15</td>\n",
       "      <td>71792584</td>\n",
       "      <td>71792805</td>\n",
       "      <td>chr15</td>\n",
       "      <td>71792584</td>\n",
       "      <td>71792805</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr15:75810000-75810429</th>\n",
       "      <td>chr15:75810000-75810429</td>\n",
       "      <td>chr15:75810000-75810429</td>\n",
       "      <td>15</td>\n",
       "      <td>75810000</td>\n",
       "      <td>75810429</td>\n",
       "      <td>chr15</td>\n",
       "      <td>75810000</td>\n",
       "      <td>75810429</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr15:76805231-76805523</th>\n",
       "      <td>chr15:76805231-76805523</td>\n",
       "      <td>chr15:76805231-76805523</td>\n",
       "      <td>15</td>\n",
       "      <td>76805231</td>\n",
       "      <td>76805523</td>\n",
       "      <td>chr15</td>\n",
       "      <td>76805231</td>\n",
       "      <td>76805523</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chr15:77885067-77885439</th>\n",
       "      <td>chr15:77885067-77885439</td>\n",
       "      <td>chr15:77885067-77885439</td>\n",
       "      <td>15</td>\n",
       "      <td>77885067</td>\n",
       "      <td>77885439</td>\n",
       "      <td>chr15</td>\n",
       "      <td>77885067</td>\n",
       "      <td>77885439</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56549 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 0                  features  \\\n",
       "chr2:148881654-148881927  chr2:148881654-148881927  chr2:148881654-148881927   \n",
       "chr8:99495798-99496023      chr8:99495798-99496023    chr8:99495798-99496023   \n",
       "chr22:21735444-21736416    chr22:21735444-21736416   chr22:21735444-21736416   \n",
       "chr3:28241304-28242015      chr3:28241304-28242015    chr3:28241304-28242015   \n",
       "chr5:154627638-154628386  chr5:154627638-154628386  chr5:154627638-154628386   \n",
       "...                                            ...                       ...   \n",
       "chr15:71448618-71449096    chr15:71448618-71449096   chr15:71448618-71449096   \n",
       "chr15:71792584-71792805    chr15:71792584-71792805   chr15:71792584-71792805   \n",
       "chr15:75810000-75810429    chr15:75810000-75810429   chr15:75810000-75810429   \n",
       "chr15:76805231-76805523    chr15:76805231-76805523   chr15:76805231-76805523   \n",
       "chr15:77885067-77885439    chr15:77885067-77885439   chr15:77885067-77885439   \n",
       "\n",
       "                         chr_no      start        end Chromosome      Start  \\\n",
       "chr2:148881654-148881927      2  148881654  148881927       chr2  148881654   \n",
       "chr8:99495798-99496023        8   99495798   99496023       chr8   99495798   \n",
       "chr22:21735444-21736416      22   21735444   21736416      chr22   21735444   \n",
       "chr3:28241304-28242015        3   28241304   28242015       chr3   28241304   \n",
       "chr5:154627638-154628386      5  154627638  154628386       chr5  154627638   \n",
       "...                         ...        ...        ...        ...        ...   \n",
       "chr15:71448618-71449096      15   71448618   71449096      chr15   71448618   \n",
       "chr15:71792584-71792805      15   71792584   71792805      chr15   71792584   \n",
       "chr15:75810000-75810429      15   75810000   75810429      chr15   75810000   \n",
       "chr15:76805231-76805523      15   76805231   76805523      chr15   76805231   \n",
       "chr15:77885067-77885439      15   77885067   77885439      chr15   77885067   \n",
       "\n",
       "                                End  n_cells  \n",
       "chr2:148881654-148881927  148881927       93  \n",
       "chr8:99495798-99496023     99496023       60  \n",
       "chr22:21735444-21736416    21736416       36  \n",
       "chr3:28241304-28242015     28242015       19  \n",
       "chr5:154627638-154628386  154628386       22  \n",
       "...                             ...      ...  \n",
       "chr15:71448618-71449096    71449096        3  \n",
       "chr15:71792584-71792805    71792805        7  \n",
       "chr15:75810000-75810429    75810429        3  \n",
       "chr15:76805231-76805523    76805523        7  \n",
       "chr15:77885067-77885439    77885439        7  \n",
       "\n",
       "[56549 rows x 9 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac_adata.var"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "f03b72e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BJ\n",
      "Pre-processing scATAC-seq...\n",
      "AnnData object with n_obs × n_vars = 258 × 56549\n",
      "    obs: 'label', 'n_genes', 'orig.ident'\n",
      "    var: '0', 'features', 'chr_no', 'start', 'end', 'Chromosome', 'Start', 'End', 'n_cells'\n",
      "Pre-processing scRNA-seq...\n",
      "AnnData object with n_obs × n_vars = 258 × 8326\n",
      "    obs: '0', 'orig.ident', 'label', 'n_genes'\n",
      "    var: '0', 'n_cells', 'features'\n",
      "Adding start site to meta file...\n",
      "Processing multimodal\n",
      "CRE-gene relationship\n",
      "Identifying candidate pairs within 100000.0 bp\n",
      "67784\n",
      "Running quadratic program... done.\n",
      "Cell with highest PDGFRA expression: ('CACTTTCGGCAT', 8)\n",
      "Expression level: [6.158457]\n",
      "WARNING: Trying to run `tl.dpt` without prior call of `tl.diffmap`. Falling back to `tl.diffmap` with default parameters.\n",
      "Epoch 0 (1.73 seconds): train loss 260.71\n",
      "Epoch Inference (5.50 seconds): test loss 257.98\n",
      "Total Time: 7.472091436386108 seconds\n",
      "Folder created at: ./results_BJ/CRE_gene\n",
      "H1\n",
      "Pre-processing scATAC-seq...\n",
      "AnnData object with n_obs × n_vars = 309 × 63567\n",
      "    obs: 'label', 'n_genes', 'orig.ident'\n",
      "    var: '0', 'features', 'chr_no', 'start', 'end', 'Chromosome', 'Start', 'End', 'n_cells'\n",
      "Pre-processing scRNA-seq...\n",
      "AnnData object with n_obs × n_vars = 309 × 9055\n",
      "    obs: '0', 'orig.ident', 'label', 'n_genes'\n",
      "    var: '0', 'n_cells', 'features'\n",
      "Adding start site to meta file...\n",
      "Processing multimodal\n",
      "CRE-gene relationship\n",
      "Identifying candidate pairs within 100000.0 bp\n",
      "73403\n",
      "Running quadratic program... done.\n",
      "Cell with highest CD9 expression: ('AATTCACGTATG', 166)\n",
      "Expression level: [6.497154]\n",
      "WARNING: Trying to run `tl.dpt` without prior call of `tl.diffmap`. Falling back to `tl.diffmap` with default parameters.\n",
      "Epoch 0 (1.32 seconds): train loss 269.96\n",
      "Epoch Inference (5.98 seconds): test loss 266.52\n",
      "Total Time: 7.478005409240723 seconds\n",
      "Folder created at: ./results_H1/CRE_gene\n",
      "GM12878\n",
      "Pre-processing scATAC-seq...\n",
      "AnnData object with n_obs × n_vars = 116 × 17461\n",
      "    obs: 'label', 'n_genes', 'orig.ident'\n",
      "    var: '0', 'features', 'chr_no', 'start', 'end', 'Chromosome', 'Start', 'End', 'n_cells'\n",
      "Pre-processing scRNA-seq...\n",
      "AnnData object with n_obs × n_vars = 116 × 3830\n",
      "    obs: '0', 'orig.ident', 'label', 'n_genes'\n",
      "    var: '0', 'n_cells', 'features'\n",
      "Adding start site to meta file...\n",
      "Processing multimodal\n",
      "CRE-gene relationship\n",
      "Identifying candidate pairs within 100000.0 bp\n",
      "12647\n",
      "Running quadratic program... done.\n",
      "Cell with highest CD38 expression: ('TCGCATCGCTGC', 62)\n",
      "Expression level: [5.622146]\n",
      "WARNING: Trying to run `tl.dpt` without prior call of `tl.diffmap`. Falling back to `tl.diffmap` with default parameters.\n",
      "Epoch 0 (0.18 seconds): train loss 67.83\n",
      "Epoch Inference (1.00 seconds): test loss 67.44\n",
      "Total Time: 1.2307779788970947 seconds\n",
      "Folder created at: ./results_GM12878/CRE_gene\n",
      "K562\n",
      "Pre-processing scATAC-seq...\n",
      "AnnData object with n_obs × n_vars = 150 × 14922\n",
      "    obs: 'label', 'n_genes', 'orig.ident'\n",
      "    var: '0', 'features', 'chr_no', 'start', 'end', 'Chromosome', 'Start', 'End', 'n_cells'\n",
      "Pre-processing scRNA-seq...\n",
      "AnnData object with n_obs × n_vars = 150 × 6365\n",
      "    obs: '0', 'orig.ident', 'label', 'n_genes'\n",
      "    var: '0', 'n_cells', 'features'\n",
      "Adding start site to meta file...\n",
      "Processing multimodal\n",
      "CRE-gene relationship\n",
      "Identifying candidate pairs within 100000.0 bp\n",
      "17467\n",
      "Running quadratic program... done.\n",
      "Gene CD38 not found in the dataset.\n",
      "WARNING: Trying to run `tl.dpt` without prior call of `tl.diffmap`. Falling back to `tl.diffmap` with default parameters.\n",
      "Epoch 0 (0.26 seconds): train loss 79.11\n",
      "Epoch Inference (1.40 seconds): test loss 78.32\n",
      "Total Time: 1.7470595836639404 seconds\n",
      "Folder created at: ./results_K562/CRE_gene\n"
     ]
    }
   ],
   "source": [
    "for celltype in celltypes:\n",
    "    print(celltype)\n",
    "    if celltype == 'BJ' :\n",
    "        marker_gene = 'PDGFRA' #fibroblast\n",
    "    elif celltype == 'H1':\n",
    "        marker_gene = 'CD9'\n",
    "    elif celltype == 'GM12878':\n",
    "        marker_gene = 'CD38'\n",
    "    elif celltype =='K562':\n",
    "        marker_gene = 'CD38'\n",
    "    atac_adata, rna_adata = preprocessing_adata_celltype(atac_anndata,rna_anndata,gene_file=gene_file,celltype = celltype)\n",
    "    \n",
    "    print(\"Processing multimodal\")\n",
    "    preprocess_multimodal(rna_adata,atac_adata)\n",
    "    print(\"CRE-gene relationship\")\n",
    "    folder_path = f\"./results_{celltype}/CRE_gene\"\n",
    "\n",
    "    run_gridot_cell_population(rna_adata,atac_adata,schema_reference='rna',marker_gene = marker_gene,distance_threshold=distance_threshold,device = device,folder_path = folder_path)\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gridot",
   "language": "python",
   "name": "gridot"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
